File: src/utils/signature.js
const crypto = require('crypto');

/**
 * Generate signature untuk Omega Tronik API
 * Berdasarkan dokumentasi: 
 * $sign = str_replace('/', '_', str_replace('+', '-' , rtrim(base64_encode(sha1($str, true)), '=')));
 */
class SignatureGenerator {
  
  /**
   * Generate signature untuk transaksi
   * @param {string} memberId - Member ID Omega Tronik
   * @param {string} product - Kode produk
   * @param {string} dest - Nomor tujuan
   * @param {string} refID - Reference/Transaction ID
   * @param {string} pin - PIN transaksi
   * @param {string} password - Password transaksi
   * @returns {string} Signature yang sudah di-encode
   */
  static forTransaction(memberId, product, dest, refID, pin, password) {
    const str = `OtomaX|${memberId}|${product}|${dest}|${refID}|${pin}|${password}`;
    return this.generate(str);
  }
  
  /**
   * Generate signature untuk cek balance
   * @param {string} memberId - Member ID Omega Tronik
   * @param {string} pin - PIN transaksi
   * @param {string} password - Password transaksi
   * @returns {string} Signature yang sudah di-encode
   */
  static forCheckBalance(memberId, pin, password) {
    const str = `OtomaX|CheckBalance|${memberId}|${pin}|${password}`;
    return this.generate(str);
  }
  
  /**
   * Generate signature untuk deposit ticket
   * @param {string} memberId - Member ID Omega Tronik
   * @param {string} pin - PIN transaksi
   * @param {string} password - Password transaksi
   * @param {number} amount - Jumlah deposit
   * @returns {string} Signature yang sudah di-encode
   */
  static forDeposit(memberId, pin, password, amount) {
    const str = `OtomaX|ticket|${memberId}|${pin}|${password}|${amount}`;
    return this.generate(str);
  }
  
  /**
   * Core function untuk generate signature
   * @param {string} str - String yang akan di-hash
   * @returns {string} Signature hasil encoding
   */
  static generate(str) {
    // SHA1 hash (binary)
    const hash = crypto.createHash('sha1').update(str).digest();
    
    // Base64 encode
    let base64 = hash.toString('base64');
    
    // Remove trailing '='
    base64 = base64.replace(/=+$/, '');
    
    // Replace '+' with '-' and '/' with '_'
    const sign = base64.replace(/\+/g, '-').replace(/\//g, '_');
    
    return sign;
  }
}

module.exports = SignatureGenerator;

File: utils/signature.py
import hashlib
import base64


def generate_signature(member_id: str, pin: str, password: str) -> str:
    """
    Generate signature for Omega Tronik API authentication (Check Balance).
    
    Format: OtomaX|CheckBalance|{memberId}|{pin}|{password}
    
    Args:
        member_id: Your Omega Tronik member ID
        pin: Your transaction PIN
        password: Your API password
        
    Returns:
        str: Base64 encoded SHA1 signature
    """
    signature_string = f"OtomaX|CheckBalance|{member_id}|{pin}|{password}"
    
    # Generate SHA1 hash
    sha1_hash = hashlib.sha1(signature_string.encode()).digest()
    
    # Base64 encode
    signature = base64.b64encode(sha1_hash).decode()
    
    # Remove trailing '='
    signature = signature.rstrip('=')
    
    # Replace '/' with '_' and '+' with '-'
    signature = signature.replace('/', '_').replace('+', '-')
    
    return signature


def generate_order_signature(member_id: str, pin: str, password: str, 
                            destination: str, product_code: str) -> str:
    """
    Generate signature for order transaction.
    
    Format: OtomaX|{memberId}|{product}|{dest}|{ref}|pin|{password}
    
    Args:
        member_id: Your Omega Tronik member ID
        pin: Your transaction PIN
        password: Your API password
        destination: Destination number
        product_code: Product code to order
        
    Returns:
        str: Base64 encoded SHA1 signature
    """
    # Generate reference ID (can be timestamp or random)
    import time
    ref_id = str(int(time.time()))
    
    signature_string = f"OtomaX|{member_id}|{product_code}|{destination}|{ref_id}|{pin}|{password}"
    
    # Generate SHA1 hash
    sha1_hash = hashlib.sha1(signature_string.encode()).digest()
    
    # Base64 encode
    signature = base64.b64encode(sha1_hash).decode()
    
    # Remove trailing '='
    signature = signature.rstrip('=')
    
    # Replace '/' with '_' and '+' with '-'
    signature = signature.replace('/', '_').replace('+', '-')
    
    return signature
